ssl/statem/statem_clnt.c

```diff
@@ -821,7 +821,10 @@ int tls_construct_client_hello(SSL *s)
         goto err;
     }
 
-    if ((sess == NULL) || (sess->ssl_version != s->version) ||
+    if ((sess == NULL) ||
+        /* Only destroy session which version is bigger than maximum supported.
+         */
+        ssl_version_cmp(s, sess->ssl_version, s->version) > 0 ||
```

Comment:
I think this scenario will actually work just fine. As far as I see it, when `TLSv1_1_client_method` - the `ssl_choose_client_version` will leave `s->version` untouched (and equal to `ctx->version`), and session version check below will fail, unless the session was created just for this connection. Does it make sense?